File: flapypan-blog-api/src/main/java/top/flapypan/blog/controller/ArticleController.java
package top.flapypan.blog.controller;

import cn.dev33.satoken.annotation.SaIgnore;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Positive;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.util.StringUtils;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;
import top.flapypan.blog.common.RestResult;
import top.flapypan.blog.dto.ArticleDTO;
import top.flapypan.blog.dto.ArticleSimpleDTO;
import top.flapypan.blog.service.ArticleService;
import top.flapypan.blog.vo.ArticleAddRequest;
import top.flapypan.blog.vo.ArticleGroupByYear;
import top.flapypan.blog.vo.ArticleUpdateRequest;

import java.util.List;

@Slf4j
@RequiredArgsConstructor
@RestController
@RequestMapping("/article")
public class ArticleController {

    private final ArticleService articleService;

    @SaIgnore
    @GetMapping
    public RestResult<Page<ArticleSimpleDTO>> getPage(String keyword, Pageable pageable) {
        Page<ArticleSimpleDTO> page;
        if (StringUtils.hasText(keyword)) {
            page = articleService.searchByKeyword(keyword, pageable);
        } else {
            page = articleService.getPage(pageable);
        }
        return RestResult.success(page);
    }

    @SaIgnore
    @GetMapping("/group-by/year")
    public RestResult<List<ArticleGroupByYear>> getGroupByYear() {
        return RestResult.success(articleService.getGroupByYear());
    }

    @SaIgnore
    @Validated
    @GetMapping("/{path}")
    public RestResult<ArticleDTO> getByPath(@PathVariable @Pattern(regexp = "^[a-z0-9:@._-]+$") String path) {
        ArticleDTO articleDTO = articleService.getByPath(path);
        return RestResult.success(articleDTO);
    }

    @PostMapping
    public RestResult<String> add(@RequestBody @Validated ArticleAddRequest articleAddRequest) {
        ArticleDTO articleDTO = articleService.add(articleAddRequest);
        return RestResult.success(articleDTO.getPath());
    }

    @PutMapping
    public RestResult<String> update(@RequestBody @Validated ArticleUpdateRequest articleUpdateRequest) {
        ArticleDTO articleDTO = articleService.update(articleUpdateRequest);
        return RestResult.success(articleDTO.getPath());
    }

    @Validated
    @DeleteMapping("/{id}")
    public RestResult<Void> delete(@PathVariable @Positive Long id) {
        articleService.delete(id);
        return RestResult.success();
    }

}

File: flapypan-blog-api/src/main/kotlin/top/flapypan/blog/controller/ArticleController.kt
package top.flapypan.blog.controller

import jakarta.validation.constraints.Pattern
import jakarta.validation.constraints.Positive
import org.springframework.data.domain.Pageable
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import top.flapypan.blog.common.restOk
import top.flapypan.blog.service.ArticleService
import top.flapypan.blog.vo.ArticleAddRequest
import top.flapypan.blog.vo.ArticleInfo
import top.flapypan.blog.vo.ArticleUpdateRequest

/**
 * 文章相关接口
 */
@RestController
@RequestMapping("/article")
class ArticleController(
    private val articleService: ArticleService
) {

    /**
     * 获取分页文章信息
     */
    @GetMapping
    fun getPage(keyword: String?, pageable: Pageable) =
        (keyword?.takeIf(String::isNotBlank)?.let {
            articleService.searchByKeyword(it, pageable)
        } ?: articleService.getPage(pageable)).map(::ArticleInfo).restOk()

    /**
     * 获取所有文章信息，并根据年份分组
     */
    @GetMapping("/group-by/year")
    fun getGroupByYear() = articleService.groupByYear().restOk()

    /**
     * 通过路径获取文章内容
     */
    @Validated
    @GetMapping("/{path}")
    fun getByPath(@PathVariable @Pattern(regexp = "^[a-z0-9:@._-]+$") path: String) =
        articleService.getByPath(path).restOk()

    /**
     * 添加文章
     */
    @PostMapping
    fun add(@RequestBody @Validated articleAddRequest: ArticleAddRequest) =
        articleService.add(articleAddRequest).restOk()

    /**
     * 修改文章
     */
    @PutMapping
    fun update(@RequestBody @Validated articleUpdateRequest: ArticleUpdateRequest) =
        articleService.update(articleUpdateRequest).restOk()

    /**
     * 删除文章
     */
    @Validated
    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: @Positive Long) =
        articleService.delete(id).restOk()
}
